lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,880 @@
+#![allow(dead_code)]
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Read, Write};
+use std::mem::size_of_val;
+use std::path::{Path, PathBuf};
+
+use bitvec::prelude::BitSlice;
+use bitvec::vec::BitVec;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use common::types::PointOffsetType;
+use memory::mmap_ops::{create_and_ensure_length, open_write_mmap};
+use uuid::Uuid;
+
+use crate::common::mmap_bitslice_buffered_update_wrapper::MmapBitSliceBufferedUpdateWrapper;
+use crate::common::mmap_slice_buffered_update_wrapper::MmapSliceBufferedUpdateWrapper;
+use crate::common::mmap_type::{MmapBitSlice, MmapSlice};
+use crate::common::operation_error::{OperationError, OperationResult};
+use crate::common::Flusher;
+use crate::id_tracker::IdTracker;
+use crate::types::{ExtendedPointId, PointIdType, SeqNumberType};
+
+pub const DELETED_FILE_NAME: &str = "id_tracker.deleted";
+pub const MAPPINGS_FILE_NAME: &str = "id_tracker.mappings";
+pub const VERSION_MAPPING_FILE_NAME: &str = "id_tracker.versions";
+
+#[derive(Debug)]
+pub struct ImmutableIdTracker {
+    path: PathBuf,
+
+    deleted: BitVec,
+    deleted_wrapper: MmapBitSliceBufferedUpdateWrapper,
+
+    internal_to_version: Vec<SeqNumberType>,
+    internal_to_version_wrapper: MmapSliceBufferedUpdateWrapper<SeqNumberType>,
+
+    mappings: PointMappings,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct PointMappings {
+    pub(crate) internal_to_external: Vec<PointIdType>,
+
+    // Having two separate maps allows us iterating only over one type at a time without having to filter.
+    pub(crate) external_to_internal_num: BTreeMap<u64, PointOffsetType>,
+    pub(crate) external_to_internal_uuid: BTreeMap<Uuid, PointOffsetType>,
+}
+
+/// Used endianness for storing PointMapping-files.
+type FileEndianess = LittleEndian;
+
+impl PointMappings {
+    const EXTERNAL_ID_NUMBER_BYTE: u8 = 0;
+    const EXTERNAL_ID_UUID_BYTE: u8 = 1;
+
+    /// Loads a `PointMappings` from the given reader. Applies an optional filter of deleted items
+    /// to prevent allocating unneeded data.
+    pub fn load<R: Read>(mut reader: R, filter: Option<&BitSlice>) -> OperationResult<Self> {
+        // Deserialize the header
+        let len = reader.read_u64::<FileEndianess>()? as usize;
+
+        let mut internal_to_external = Vec::with_capacity(len);
+        let mut external_to_internal_num: BTreeMap<u64, PointOffsetType> = BTreeMap::new();
+        let mut external_to_internal_uuid: BTreeMap<Uuid, PointOffsetType> = BTreeMap::new();
+
+        // Deserialize the list entries
+        for i in 0..len {
+            let (internal_id, external_id) = Self::read_entry(&mut reader)?;
+
+            // Need to push this regardless of point deletion as the vecs index represents the internal id
+            // which would become wrong if we leave out entries.
+            internal_to_external.push(external_id);
```

Comment:
AFAIK, `SimpleIdTracker` use `u64::MAX` to mark deleted internal-to-external mappings, so we might want to handle this similarly for `ImmutableIdTracker`?